pyasic/miners/base.py

```diff
@@ -540,3 +545,115 @@ def __init__(self, ip: str) -> None:
 
 
 AnyMiner = TypeVar("AnyMiner", bound=BaseMiner)
+
+
+class FirmwareManager:
+    class FirmwareManager:
+        def __init__(self, remote_server_url: str):
+            """
+            Initialize a FirmwareManager instance.
+
+            Args:
+                remote_server_url (str): The URL of the remote server to fetch firmware information.
+            """
+            self.remote_server_url = remote_server_url
+            self.version_extractors = {}
+
+            # Register version extractor for braiins_os
+            self.register_version_extractor("braiins_os", self.extract_braiins_os_version)
+
+    def extract_braiins_os_version(self, firmware_file: Path) -> str:
+        """
+        Extract the firmware version from the filename for braiins_os miners.
+
+        Args:
+            firmware_file (Path): The firmware file to extract the version from.
+
+        Returns:
+            str: The extracted firmware version.
+
+        Raises:
+            ValueError: If the version is not found in the filename.
+        """
+        match = re.search(r"firmware_v(\d+\.\d+\.\d+)\.tar\.gz", firmware_file.name)
```

Comment:
Very close here, this firmware feeds server is a bit complex in that you have to match subdirectories, and also some miners expect a specific type of update.  

If you take a look at feeds.braiins-os.com you'll see what I mean, there are am1_s9 upgrades, am2_x17 upgrades, and am3_bbb upgrades, and they are all for different types.

Also, firmware upgrades are just .tar, not .tar.gz.